lib/Twig/Extension/Core.php

```diff
@@ -821,7 +821,27 @@ function _twig_default_filter($value, $default = '')
 function twig_get_array_keys_filter($array)
 {
     if ($array instanceof Traversable) {
-        return array_keys(iterator_to_array($array));
+        while ($array instanceof IteratorAggregate) {
+            $array = $array->getIterator();
+        }
+
+        if ($array instanceof Iterator) {
+            $keys = array();
+            $array->rewind();
+            while ($array->valid()) {
+                $keys[] = $array->key();
+                $array->next();
+            }
+
+            return $keys;
+        }
```

Comment:
Awesome, thanks for all the work :)
I'm looking for a real-world cases that are opensource.
However such cases are plenty and easy to think of. When iterate data:
- with fix length/size files data (for example binary files)
- with delimiters ('csv', 'json')
- such as database result sets based on identifiers
- such as cached identifiers with the values in some other source (like a database)

Whenever you want filter by key (or when only interested in the key for whatever reason) there is a clear advantage by not fetching the value (for example hydrating an item by doctrine2 is pretty expensive especially if you are not going to need it).

For example:
small gain [here](https://github.com/doctrine/doctrine2/blob/24377156b41a5fca37b26579c28f824f779d41ef/lib/Doctrine/ORM/Query/TreeWalkerChainIterator.php#L64) and even for more gain after some changes in classes like [this](https://github.com/doctrine/doctrine2/blob/cd11723e6310b5973f50e7d9819c58916721d290/lib/Doctrine/ORM/Internal/Hydration/IterableResult.php)